### PR TITLE
fix(streaming): correct `JoinRow` serialization

### LINKED ISSUE #2649: Serialization of JoinRow does not handle null value
`JoinRow` may contain null value. But it was simply skipped.

### PR DESCRIPTION
**Pull request by yuhao-su**
## What's changed and what's your intention?

- serialize pk of `JoinHashMap` with value encoding.
- correct the serialization of `JoinRow` with a null value.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close #2649

### CODE REVIEW

File: src/common/src/util/value_encoding/mod.rs [hunk 1]
```diff
@@ -44,6 +44,28 @@ pub fn deserialize_cell(mut data: impl Buf, ty: &DataType) -> Result<Datum> {
     deserialize_value(ty, &mut data)
 }
 
+/// Serialize a datum into bytes (Not order guarantee, used in value encoding).
+pub fn serialize_datum(cell: &Datum) -> Result<Vec<u8>> {
```
Comment (WindowsXp-Beta) [inline src/common/src/util/value_encoding/mod.rs:48]: You may rename it to `serialize_cell` and change the origin one into `serialize_cell_not_null` which we used to have before my commit and we think the null tag is useless and remove it🤣.

File: src/common/src/util/value_encoding/mod.rs [hunk 2]
```diff
@@ -44,6 +44,28 @@ pub fn deserialize_cell(mut data: impl Buf, ty: &DataType) -> Result<Datum> {
...
+        buf.put_u8(1);
+        serialize_value(datum.as_scalar_ref_impl(), &mut buf)
+    } else {
+        buf.put_u8(0);
+    }
+    Ok(buf)
+}
+
+/// Deserialize bytes into a datum (Not order guarantee, used in value encoding).
+pub fn deserialize_datum(mut data: impl Buf, ty: &DataType) -> Result<Datum> {
```
Comment (WindowsXp-Beta) [inline src/common/src/util/value_encoding/mod.rs:60]: ditto

File: src/common/src/util/value_encoding/mod.rs (hunk 1 above)
Comment (yuhao-su) [reply src/common/src/util/value_encoding/mod.rs:48]: I think we should keep the current names. Current `serialize_cell` actually can serialize a datum into an empty vec, which is the expected behavior in cell based context. And the  `serialize_datum` function will never be used in cell based context, it only loyally encodes the datum.

File: src/stream/src/executor/managed_state/join/mod.rs [hunk 3]
```diff
@@ -101,17 +96,11 @@ impl JoinRowDeserializer {
     }
 
     /// Deserialize the row from a memcomparable bytes.
```
Comment (BowenXiao1999) [inline src/stream/src/executor/managed_state/join/mod.rs:98]: value-encoding bytes

File: src/common/src/util/value_encoding/mod.rs (hunk 1 above)
Comment (BowenXiao1999) [reply src/common/src/util/value_encoding/mod.rs:48]: Not sure what's the diff between `serialize_cell` and `serialize_datum` in your understanding? I think this two function is hard to tell the difference from name. May also add some function comments?

File: src/common/src/array/data_chunk_iter.rs [hunk 4]
```diff
@@ -254,6 +257,17 @@ impl Row {
...
 
+    /// Serialize the row into a value encode bytes.
+    ///
+    /// All values are nullable. Each value will have 1 extra byte to indicate whether it is null.
+    pub fn value_encode(&self) -> RwResult<Vec<u8>> {
+        let mut vec = vec![];
+        for v in &self.0 {
+            vec.extend_from_slice(&serialize_datum(v)?);
+        }
+        Ok(vec)
```
Comment (BowenXiao1999) [inline src/common/src/array/data_chunk_iter.rs:268]: Rewrite to self.0.iter.map.collect?

File: src/common/src/array/data_chunk_iter.rs [hunk 5]
```diff
@@ -326,6 +340,16 @@ impl RowDeserializer {
         let datum = deserialize_datum_from(&self.data_types[datum_idx], &mut deserializer)?;
         Ok(datum)
     }
+
+    /// Deserialize the row from a value encoding bytes.
+    pub fn value_decode(&self, mut data: impl Buf) -> RwResult<Row> {
+        let mut values = vec![];
+        values.reserve(self.data_types.len());
+        for ty in &self.data_types {
+            values.push(deserialize_datum(&mut data, ty)?);
```
Comment (BowenXiao1999) [inline src/common/src/array/data_chunk_iter.rs:349]: ditto

**Review by BowenXiao1999 — COMMENTED** [review]
Rest LGTM

File: src/common/src/array/data_chunk_iter.rs (hunk 4 above)
Comment (yuhao-su) [reply src/common/src/array/data_chunk_iter.rs:268]: Should be flat map :>

File: src/common/src/array/data_chunk_iter.rs (hunk 5 above)
Comment (yuhao-su) [reply src/common/src/array/data_chunk_iter.rs:349]: `deserialize_datum` has to returns a `Result`. Can't return `Result` in a map function.

File: src/common/src/array/data_chunk_iter.rs [hunk 6]
```diff
@@ -326,6 +340,16 @@ impl RowDeserializer {
         let datum = deserialize_datum_from(&self.data_types[datum_idx], &mut deserializer)?;
         Ok(datum)
     }
+
+    /// Deserialize the row from a value encoding bytes.
+    pub fn value_decode(&self, mut data: impl Buf) -> RwResult<Row> {
+        let mut values = vec![];
+        values.reserve(self.data_types.len());
```
Comment (BugenZhao) [inline src/common/src/array/data_chunk_iter.rs:347]:
```suggestion
        let mut values = Vec::with_capacity(self.data_types.len());
```

File: src/common/src/array/data_chunk_iter.rs [hunk 7]
```diff
@@ -254,6 +257,17 @@ impl Row {
         Ok(serializer.into_inner())
     }
 
+    /// Serialize the row into a value encode bytes.
+    ///
+    /// All values are nullable. Each value will have 1 extra byte to indicate whether it is null.
+    pub fn value_encode(&self) -> RwResult<Vec<u8>> {
+        let mut vec = vec![];
+        for v in &self.0 {
+            vec.extend_from_slice(&serialize_datum(v)?);
```
Comment (BugenZhao) [inline src/common/src/array/data_chunk_iter.rs:266]:
```suggestion
            vec.extend(serialize_datum(v)?);
```

**Review by BowenXiao1999 — APPROVED** [review]

**Pull request merged**